document/core/exec/instructions.rst

```diff
@@ -182,6 +182,527 @@ Where the underlying operators are non-deterministic, because they may return on
+
+2. Pop the value :math:`\V128.\VCONST~c_1` from the stack.
+
+3. Let :math:`n` be the integer :math:`\dim(\shape)`.
+
+4. Let :math:`i_1^n` be the sequence :math:`\lanes_{\shape}(c_1)`
+
+5. Let :math:`i` be the result of computing :math:`\bigwedge(i_1 \neq 0)^n`.
+
+6. Push the value :math:`\I32.\CONST~i` onto the stack.
+
+
+.. math::
+   \begin{array}{l}
+   \begin{array}{lcl@{\qquad}l}
+   (\V128\K{.}\VCONST~c_1)~\shape\K{.}\ALLTRUE &\stepto& (\I32\K{.}\CONST~i)
+   \end{array}
+   \\ \qquad
+     \begin{array}[t]{@{}r@{~}l@{}}
+     (\iff & n = \dim(\shape) \\
+     \wedge & i_1^n = \lanes_{\shape}(c) \\
+     \wedge & i = \bigwedge(i_1 \neq 0)^n
+     \end{array}
+   \end{array}
+
+
+.. _exec-simd-anytrue:
+
+:math:`\shape\K{.}\ANYTRUE`
+...........................
+
+1. Assert: due to :ref:`validation <valid-vitestop>`, a value of :ref:`value type <syntax-valtype>` |V128| is on the top of the stack.
+
+2. Pop the value :math:`\V128.\VCONST~c_1` from the stack.
+
+3. Let :math:`i` be the result of computing :math:`\ine_{\K{128}}(c_1, 0)`.
+
+4. Push the value :math:`\I32.\CONST~i` onto the stack.
+
+.. math::
+   \begin{array}{lcl@{\qquad}l}
+   (\V128\K{.}\VCONST~c_1)~\shape\K{.}\ANYTRUE &\stepto& (\I32\K{.}\CONST~i)
+     & (\iff i = \ine_{\K{128}}(c_1, 0)) \\
+   \end{array}
+
+
+.. _exec-simd-bitmask:
+
+:math:`t\K{x}N\K{.}\BITMASK`
+............................
+
+1. Assert: due to :ref:`validation <valid-vitestop>`, a value of :ref:`value type <syntax-valtype>` |V128| is on the top of the stack.
+
+2. Pop the value :math:`\V128.\VCONST~c_1` from the stack.
+
+3. Let :math:`i_1^N` be the sequence :math:`\lanes_{t\K{x}N}(c)`.
+
+4. Let :math:`B` be the :ref:`bit width <syntax-valtype>` :math:`|t|` of :ref:`value type <syntax-valtype>` :math:`t`.
+
+5. Let :math:`i_2^N` be the sequence as a result of computing :math:`\ilts_{B}(i_1^N, 0^N)`.
+
+6. Let :math:`c` be the integer :math:`\ibits_{32}^{-1}(i_2^N~0^{32-N})`.
+
+7. Push the value :math:`\I32.\CONST~c` onto the stack.
+
+.. math::
+   \begin{array}{lcl@{\qquad}l}
+   (\V128\K{.}\VCONST~c_1)~\shape\K{.}\BITMASK &\stepto& (\I32\K{.}\CONST~c)
+     & (\iff c = \ibits_{32}^{-1}(\ilts_{B}(\lanes_{t\K{x}N}(c), 0^N)))
+     \\
+   \end{array}
+
+
+.. _exec-simd-narrow:
+
+:math:`t_2\K{x}N\K{.}\NARROW\_t_1\K{x}M\_\sx`
+.............................................
+
+1. Assert: due to :ref:`validation <valid-vitestop>`, two values of :ref:`value type <syntax-valtype>` |V128| are on the top of the stack.
+
+2. Pop the value :math:`\V128.\VCONST~c_2` from the stack.
+
+3. Let :math:`d_2^M` be the result of computing :math:`\narrow^{\sx}_{t_1, t_2}(\lanes_{t_1\K{x}M}(c_2))`.
+
+4. Pop the value :math:`\V128.\VCONST~c_1` from the stack.
+
+5. Let :math:`d_1^M` be the result of computing :math:`\narrow^{\sx}_{t_1, t_2}(\lanes_{t_1\K{x}M}(c_1))`.
+
+6. Let :math:`c` be the result of :math:`\lanes^{-1}_{t_2\K{x}N}(d_1^M~d_2^M)`.
```

Comment:
Does this need an assertion that N = 2M, due to syntactic constraints?

Btw, I'm now confused about the inclusion of bare `\NARROW` in vbinop in the syntax conventions. I suppose that should be `\NARROW_\shape`? Similar question about `\WIDEN`.

Reply:
> Does this need an assertion that N = 2M, due to syntactic constraints?

If we only get here from valid instructions, then we don't need such an assertion. The valid instructions are already defined such that N = 2M.

> Btw, I'm now confused about the inclusion of bare \NARROW in vbinop in the syntax conventions.

What I intended is that these narrow instructions are validated the same as the other vbinop, but for execution, it "pattern-matches" to this more specific definition of narrow (`shape.vbinop` wouldn't work for narrow/widen). I'm not sure if this works though.

I guess narrow and widen probably shouldn't be included in the syntax conventions, and should be separated out in the validation?

Reply:
> If we only get here from valid instructions, then we don't need such an assertion. The valid instructions are already defined such that N = 2M.

Yup, that's why I suggested saying "due to syntactic constraints". The nature of an assertion of course is that it's always true. But the constraint is not apparent from the pattern match above, so local to this rule, it is not clear otherwise.

> > Btw, I'm now confused about the inclusion of bare \NARROW in vbinop in the syntax conventions.
> 
> What I intended is that these narrow instructions are validated the same as the other vbinop, but for execution, it "pattern-matches" to this more specific definition of narrow (`shape.vbinop` wouldn't work for narrow/widen). I'm not sure if this works though.

You can still do so, but you have to list them in vunop/vbinop with their suffixes, i.e. `\NARROW\K{\_}\shape\K{\_}\sx` and `\WIDEN\K{\_low\_}\shape\K{\_}\sx | \WIDEN\K{\_high\_}\shape\K{\_}\sx`, respectively. The same for `\VTRUNC` and `\VCONVERT`, now that I notice it.

Reply:
> Yup, that's why I suggested saying "due to syntactic constraints".

Added this assertion.

> You can still do so, but you have to list them in vunop/vbinop with their suffixes,

Got it, fixed.